### PR TITLE
Bump dependencies to latest versions

### PR DESCRIPTION
**Pull request by jelmer**
Bump dependencies to latest versions. This required some minor adjustments to APIs:

* gen_range in rand now takes a range rather than start and end (gen_range(X, Y) -> gen_range(..))
* GTK Rectangle fields are hidden now, and require accessor functions / explicit constructor
* "get_" has been dropped from the name of various Gtk functions
* "new_" has been dropped from the name of various Gtk functions
* "CheckBox::get_active" has been renamed to "CheckBox::is_active"
* various cairo functions now return a Result<(), Error> rather than ()
* GtkWindow.destroy() is now unsafe; I've just added unsafe {} around it for now, but there's presumably a better alternative that doesn't require unsafe {}

### CODE REVIEW

**Comment by jelmer** [comment]
ping @sprang

File: examples/genpng.rs [hunk 1]
```diff
@@ -65,53 +69,61 @@ fn generate_card_images(path: &str, card_width: i32, border: i32,
...
+            clap::arg!(-b --border <BORDER> "Sets the border width in pixels"),
+            clap::arg!(-w --width <WIDTH> "Sets the card width in pixels"),
+        ])
+        .get_matches();
+
+    let path = matches.get_one::<String>("DIRECTORY").unwrap();
+    let width = *matches.get_one::<i32>("WIDTH").unwrap_or(&350 as &i32);
+    let border = *matches.get_one::<i32>("BORDER").unwrap_or(&0 as &i32);
+    let render_vertically = matches.get_flag("VERTICAL");
+    let classic_colors = matches.get_flag("CLASSIC");
```
Comment (sprang) [inline examples/genpng.rs]: This doesn't seem to be handling command line arguments correctly. If I specify a directory I get a panic. Maybe it worked when you first made the changes (the version range for 'clap' in Cargo.toml is quite broad)?

File: examples/simulate.rs [hunk 2]
```diff
@@ -321,24 +321,25 @@ fn run_simulations(num_games: u64, num_threads: u64) {
...
-    let matches = clap_app!(simulate =>
-        (version: VERSION)
-        (about: "Gather statistics for simulated games of SET.")
-        (@arg GAMES: -g --games +takes_value games_help)
-        (@arg THREADS: -t --threads +takes_value "Sets number of threads")
-    ).get_matches();
-
-    let num_games = value_t!(matches, "GAMES", u64).unwrap_or(NUM_GAMES);
-    let num_threads = value_t!(matches, "THREADS", u64).unwrap_or(num_cpus::get() as u64);
+    let games_default = pretty_print(NUM_GAMES).to_string();
```
Comment (sprang) [inline examples/simulate.rs]: _games_default_ appears to be unused.

File: examples/simulate.rs [hunk 3]
```diff
@@ -321,24 +321,25 @@ fn run_simulations(num_games: u64, num_threads: u64) {
...
+        .args([
+            clap::arg!(-g --games <GAMES> "Sets number of games to simulate"),
+            clap::arg!(-t --threads <THREADS> "Sets number of threads"),
+        ])
+        .get_matches();
+
+    let num_games = *matches.get_one::<u64>("GAMES").unwrap_or(&NUM_GAMES);
+    let num_threads = *matches
+        .get_one::<usize>("THREADS")
+        .unwrap_or(&num_cpus::get());
```
Comment (sprang) [inline examples/simulate.rs]: Something doesn't seem to be working here. I can't adjust the number of games or threads from the command line.

File: src/core/graphics.rs [hunk 4]
```diff
@@ -134,118 +156,169 @@ impl ContextExt for Context {
...
-        let Rectangle { x, y, width, height } = rect;
+    fn draw_card(
+        &self,
+        card: Card,
+        rect: Rectangle,
+        label: Option<&str>,
+        scheme: ColorScheme,
+    ) -> Result<(), Error> {
+        let x = rect.height();
+        let y = rect.width();
```
Comment (sprang) [inline src/core/graphics.rs]: This introduced a rendering bug. Should be:
```
        let x = rect.x();
        let y = rect.y();
```

File: src/main.rs [hunk 5]
```diff
@@ -243,37 +248,33 @@ fn build_deck_submenu(menu_data: MenuData) -> MenuItem {
...
-	    }
-	})
-    );
+    controller
+        .borrow_mut()
+        .add_undo_observer(clone!(@weak undo, @weak redo => move |controller| {
+            undo.set_sensitive(controller.can_undo());
+            redo.set_sensitive(controller.can_redo());
+
+            if let Some(action) = controller.undo_action_name() {
```
Comment (sprang) [inline src/main.rs:266]: Indention is wonky here.

**Review by sprang — CHANGES_REQUESTED** [review]
Thanks for updating all these dependencies! Just a few issues to resolve.

**Comment by jelmer** [comment]
PTAL - I believe these should all be fixed now.

**Review by jelmer — COMMENTED** [review]
ping :)

**Comment by sprang** [comment]
Thank you! Sorry it took me a while to pay attention to this.

**Pull request merged**